### ISSUE TITLE
All tools assign instruments incorrectly for San Francisco Rush 2049

### ISSUE THREAD

**Issue by JillCrungus** [jombo23/N64-Tools#34]
When exporting MIDIs from San Francisco Rush 2049, the tool is just assigning instruments in sequential order (track 1 gets instrument 0, track 2 gets instrument 1, etc.), which definitely doesn't seem to be correct and results in a mangled sound when the game's soundfont is applied.

Either the game really assigns instruments in this way and the DLS export is broken (incredibly unlikely) or the tools are assigning instruments wrong.

**Comment by SubDrag** [comment]
This is a Factor 5 game. Did you try using Write DLS Combined banks? That should get you the more correct lookup.

**Comment by JillCrungus** [comment]
> This is a Factor 5 game. Did you try using Write DLS Combined banks? That should get you the more correct lookup.

Can you elaborate on the right usage of this? It outputs more banks and I assumed that the banks would map to tracks but none of the tracks seem to sound correct regardless of the bank used to play them.

**Comment by SubDrag** [comment]
I believe every song gets its own associated bank (basically it remaps the real numbers mapped to 0-15).  So you want to play the song with the analogous bank.

**Comment by JillCrungus** [comment]
> I believe every song gets its own associated bank (basically it remaps the real numbers mapped to 0-15). So you want to play the song with the analogous bank.

That was my assumption, but no matter which bank I use the tracks always seem to sound completely wrong.

**Comment by SubDrag** [comment]
It should match 1:1 - one DLS per midi.  If that's not sounding right, I'm not sure, maybe this game has a lookup of instruments different than the standard.

**Comment by SubDrag** [comment]
OK so I'm looking at the code...it's possible I never actually supported Factor 5's soundbanks properly.

**Comment by SubDrag** [comment]
Star Wars Rogue Squadron seems to have issues with the DLS, but they seem ok.  It might be problems with Viena/Synthfont though.  The audiobanks at least seem to have the right instruments there.  I guess Factor 5 games unfortunately aren't going to work right now.

**Comment by SubDrag** [comment]
Yeah it uses the Sound Macro list, unlike the other games which all use Layers.  Also note it's not a 1:1 of song # to group.  Sorry I don't think there'll be a solution any time soon.  For some reason the sampleId id it spits out are not the real numbers and are some kind of mapping that is unique to that game.

For example Rogue Squadron has this mapping:
https://gitlab.axiodl.com/AxioDL/amuse/-/blob/master/lib/ContainerRegistry.cpp#L650

**Comment by SubDrag** [comment]
OK I see the issue - actually all games are off.  Working on it.

**Comment by SubDrag** [comment]
OK I was accidentally not looking up by "Sfx id".  So it was the wrong samples.  Can you try the latest?  You will still have to cross-reference on your own, what soundbank goes to what midi, but it should at least be music samples now.

**Comment by SubDrag** [comment]
If this is actually correct, let me know if you figure out what maps to what, and can default that game to work that way.

**Comment by JillCrungus** [comment]
> If this is actually correct, let me know if you figure out what maps to what, and can default that game to work that way. Might line up 1:1 though on numbers in this one.

The soundbank tool shows a warning about there not being a layer offset when exporting the soundbanks which I assume is normal given what you've mentioned.

The results are definitely sounding better in the latest release. The banks seem to contain more appropriate samples now, but things still seem amiss. While the samples are clearly _better_, it's still the case that things aren't matching up at all and in some cases I've even hunted through every patch in every bank and been unable to find patches that correspond to certain instruments in some tracks, so unless I missed something, something still seems amiss.

**Comment by SubDrag** [comment]
Wait the instruments don't even exist in the sound tool? Maybe if you hone in on one song and try and sort out what it should be versus which is closest bank and what is there today... Maybe can find a pattern.

**Comment by JillCrungus** [comment]
> Wait the instruments don't even exist in the sound tool? Maybe if you hone in on one song and try and sort out what it should be versus which is closest bank and what is there today... Maybe can find a pattern.

I couldn't find certain instruments in the exported banks is what I meant. I've been testing using a specific track and went through every bank trying to get a specific instrument to match but regardless of which bank or patch was used I couldn't get a match. It is entirely possible I just missed it I suppose.

**Comment by SubDrag** [comment]
Try using the N64 sound tool. It has all samples. Not just those by songs.

**Comment by JillCrungus** [comment]
> Try using the N64 sound tool. It has all samples. Not just those by songs.

I exported all the samples with the sound tool and was able to locate the sample for the channel I'm testing with with them (named "BANK_00_INSTR_0015_SND_0000"). I'm almost certain that it wasn't in any of the soundbanks when I checked them, though it was a bit of a tedious process so it may be in there and I just missed it.

**Comment by SubDrag** [comment]
Can you pick one song and identify all the correct sound indexes versus midi track? I might just be off slightly... But I won't know till know the correct answer.

That one you found that was missing not sure what song or track you meant it was for?

**Comment by JillCrungus** [comment]
> Can you pick one song and identify all the correct sound indexes versus midi track? I might just be off slightly... But I won't know till know the correct answer.
> 
> That one you found that was missing not sure what song or track you meant it was for?

I can make an attempt, I'm not certain if I'll be able to pin down *every* sound in the track but I can try. I've been testing using the US release of the game and the track I've been testing with is the one exported by the soundbank tool as "00000007 003218B0". It's supposed to sound like this:
https://youtu.be/HXIElifH98c

The sample I mentioned is intended for the start of MIDI track 8 in that song (the drum from about 24 seconds into that video) 

Something I've just realised that's worth nothing, it seems like that track is intended to switch instruments as the latter end of it is almost certainly not for percussion but the actual MIDI doesn't change patch (I'm not super knowledgeable on how it'd work internally here, none of the notes in the latter end are shared with the ones used for the drum at the start so maybe this would be a case where the instrument would just have different samples assigned for the notes used in that section, hence no actual patch change?). If instruments can change that might make it even more difficult to pin down all the samples.

**Comment by SubDrag** [comment]
Yeah an attempt would help me see if there is a pattern. Is it only the drum track always off? Those are a bit different for Factor 5.  Against which combined soundbank.

**Comment by JillCrungus** [comment]
> Yeah an attempt would help me see if there is a pattern. Is it only the drum track always off? Those are a bit different for Factor 5. Against which combined soundbank.

Almost all of the tracks are off when using the soundbanks, so it's difficult to tell which bank is even correct (especially since there seems to be missing instruments.)

Using soundbank 0A with this track, MIDI track 11 is actually correct though all the others are still wrong so I can't even tell if it's the right bank or if this is just another bank that coincidentally has the same instrument on that track. (MIDI track 12 should be using the same instrument as track 11 at the start of the track)

**Comment by SubDrag** [comment]
Thanks that was really helpful.  OK, I think getting closer. It seems the instruments were assigned wrong...it was using the channel, instead of the instrument.  I'm wondering now if the key base they give you is not centered the same?  Can you try grabbing again and see what happens now.

**Comment by JillCrungus** [comment]
> Thanks that was really helpful. OK, I think getting closer. It seems the instruments were assigned wrong...it was using the channel, instead of the instrument. I'm wondering now if the key base they give you is not centered the same? Can you try grabbing again and see what happens now.

Definitely getting there. More of the instruments are correct now. Though, a few tracks are still off (the aforementioned drum sound - MIDI track 8, seems to be one of them.

The pitches are also definitely wrong, though I'm not sure that if it's consistent how much they're off by or not at the moment.

**Comment by SubDrag** [comment]
I found one more problem a moment ago. Just to be safe regrab and see instruments if they're OK now. Hmm drum track may be special, but maybe let's worry about the others first. Are all key base off by a consistent amount?

**Comment by SubDrag** [comment]
OK I found out I was using the wrong sampling rate on instruments...can you redownload and try again.  It actually sounds like music now I think.

**Comment by JillCrungus** [comment]
> OK I found out I was using the wrong sampling rate on instruments...can you redownload and try again. It actually sounds like music now I think.

I'm not really hearing a difference between the version I just downloaded and the last one.

Oh, I will note that if I tick the "1/2 sampling rate" box in the soundbank tool then some of the tracks sound more correct though I'm not sure they all are (it's hard to tell since some of the instruments still seem to be wrong) so maybe the sample rate is still off?

I'll try and start matching tracks to exported soundbank instruments from the sound tool some time soon

**Comment by SubDrag** [comment]
When I click the preview button it sounds pretty good?  Maybe it's the drum track - are you having your midi software treat it not as a special drum track?

**Comment by SubDrag** [comment]
If you click preview it sounds decent, I think.  The sampling rates should be better now though?   I wonder if it's the special drum track mainly causing the problems.
